### PR TITLE
#1228 multiroot interpreter display

### PR DESCRIPTION
**Pull request by DonJayamanne**
Adds multi root support for interpreter support.
I know the UX hasn't been sorted out.
I'll discuss this with John today and battle on with the last part of multi-root and that's Unit Tests

### CODE REVIEW

**Comment by brettcannon** [comment]
So is this the final merge of the overall change that has collected all the previous PRs?

File: src/client/common/interpreterInfoCache.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        return cache.has(cacheKey) ? cache.get(cacheKey) : {};
+    }
+    // tslint:disable-next-line:function-name
+    public static setPaths(resource?: Uri, pythonSettingsPath?: string, pythonInterpreterPath?: string, pythonInterpreterDirectory?: string) {
+        InterpreterInfoCache.setCacheData('pythonInterpreterDirectory', resource, pythonInterpreterDirectory);
+        InterpreterInfoCache.setCacheData('pythonInterpreterPath', resource, pythonInterpreterPath);
+        InterpreterInfoCache.setCacheData('pythonSettingsPath', resource, pythonSettingsPath);
+    }
+
+    // tslint:disable-next-line:no-any function-name
```
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/client/common/interpreterInfoCache.ts]: No it's to treat a variable as one that has no type information. (Kind of like removing type hints when a variable has had type hints defined)

File: src/client/common/interpreterInfoCache.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    public static setCustomEnvVariables(resource?: Uri, envVars?: any) {
+        // tslint:disable-next-line:no-any
+        InterpreterInfoCache.setCacheData('customEnvVariables', resource, envVars);
+    }
+    // tslint:disable-next-line:no-any function-name
+    private static setCacheData(property: keyof InterpreterCache, resource?: Uri, value?: string | any) {
```
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/client/common/interpreterInfoCache.ts]: Huh, can't understand why that was added. Yes you are right

File: src/client/common/interpreterInfoCache.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/client/common/interpreterInfoCache.ts]: Fixed

**Pull request merged**